feat(seal): detect no-prefix packages by version suffix (#10911)
aquasecurity/trivy: docs/guide/coverage/others/seal.md

Reviewer: Perhaps it makes sense to add links to the Seal docs:
- https://docs.sealsecurity.io/reference/naming-and-versioning/sp-model
- https://docs.sealsecurity.io/reference/naming-and-versioning/per-ecosystem
- https://docs.sealsecurity.io/reference/naming-and-versioning/renamed-packages

Author: Added links to sp-model, per-ecosystem, and renamed-packages in the Application Dependencies section (718a6f5).

## Application Dependencies

> Seal also provides patched versions of application dependencies with their own vulnerability advisories. Seal ships these packages under two naming schemes, and Trivy detects both. When Trivy detects a Seal-patched package by either scheme, it automatically uses Seal Security advisories for vulnerability scanning.

Reviewer: ```suggestion
Seal also provides patched versions of application dependencies, shipped under two naming schemes — [renamed](#renamed-packages) and [no-prefix](#no-prefix-packages). Trivy detects both and uses Seal Security's own advisories for vulnerability scanning instead of the upstream ones.
```

Author: Applied (kept it comma-separated rather than a dash) (a8527ff92).

| Ecosystem | Package Pattern | Example |
|-----------|----------------|---------|
| Python (pip) | `seal-*` | `seal-requests` `2.14.2+sp1` |
| Node.js (npm) | `@seal-security/*` | `@seal-security/ejs` `3.1.8-sp1` |
| Go | `sealsecurity.io/*` | `sealsecurity.io/github.com/Masterminds/goutils` `1.1.1-sp1` |
| Java (Maven) | `seal.sp*` | `seal.sp1.org.eclipse.jetty:jetty-http` |
| Ruby (RubyGems) | `seal-*` | `seal-rack` `2.0.7.0.1.sp1` |

### No-prefix packages

Some Seal packages keep their upstream (no-prefix) name and only add a patch-level version suffix. Trivy detects these by the version suffix:

| Ecosystem | Version Suffix | Example |
|-----------|----------------|---------|
| Java (Maven) | `+spN` | `org.eclipse.jetty:jetty-http` `9.4.48+sp1` |
| Python (pip) | `+spN` | `requests` `2.14.2+sp1` |
| Node.js (npm) | `-spN` | `ejs` `3.1.8-sp1` |
| Go | `-spN` | `golang.org/x/crypto` `0.26.0-sp1` |
| Ruby (RubyGems) | `.0.1.spN` | `rack` `2.0.7.0.1.sp1` |

For Maven and pip, the `+spN` suffix cannot collide with real package versions, so the match is authoritative.

> For npm, Go, and Ruby, the `-spN` / `.0.1.spN` suffix can also appear on real packages, so Trivy confirms the match by looking the package up in the Seal advisory database; if the package is not found there, it falls back to the standard ecosystem advisories.

Reviewer: I think we can move this into `note`:

```suggestion
!!! note
    For Maven and pip, the `+spN` suffix cannot collide with real package versions, so the match is authoritative.
    For npm, Go, and Ruby, the `-spN` / `.0.1.spN` suffix can also appear on real packages, so Trivy confirms the match by looking the package up in the Seal advisory database; if the package is not found there, it falls back to the standard ecosystem advisories.

```

Author: Applied (a8527ff92).

## Application Dependencies

Seal also provides patched versions of application dependencies with their own vulnerability advisories. Seal ships these packages under two naming schemes, and Trivy detects both.

> ### Renamed packages

Reviewer: Should we write that renamed packages (except Maven) also use spN suffixes for the version?

Author: Added — the Renamed table now notes that all ecosystems except Maven also carry the spN version suffix, with examples (718a6f5).

Author: On reflection, rather than duplicate the versioning rules here, I linked the Seal [renamed-packages](https://docs.sealsecurity.io/reference/naming-and-versioning/renamed-packages) doc for the full naming/versioning details (kept the table name-only). e9fb3575d.

| Java (Maven) | `seal.sp*` | `seal.sp1.org.eclipse.jetty:jetty-http` |
| Ruby (RubyGems) | `seal-*` | `seal-rack` |

### No-prefix packages

Some Seal packages keep their upstream (no-prefix) name and only add a patch-level version suffix (`+spN` / `-spN`). Trivy detects these by the version suffix:

| Ecosystem | Version Suffix | Example |
|-----------|----------------|---------|
| Java (Maven) | `+spN` | `org.eclipse.jetty:jetty-http` `9.4.48+sp1` |
| Python (pip) | `+spN` | `requests` `2.14.2+sp1` |
| Node.js (npm) | `-spN` | `ejs` `3.1.8-sp1` |
| Go | `-spN` | `golang.org/x/crypto` `0.26.0-sp1` |

> | Ruby (RubyGems) | `.spN` | `rack` `2.0.7.0.1.sp1` |

Reviewer: ```suggestion
| Ruby (RubyGems) | `.0.1.spN` | `rack` `2.0.7.0.1.sp1` |
```

Author: Applied (718a6f5).

Both public (`spN`) and private (`spNpM`) sealed versions are recognized. A private version carries an extra `pM` iteration on top of the sealed version, for example `ejs` `3.1.8-sp2p1`.

For Maven and pip, the `+spN` suffix cannot collide with real package versions, so the match is authoritative.
For npm, Go, and Ruby, the `-spN` / `.spN` suffix can also appear on real packages, so Trivy confirms the match by looking the package up in the Seal advisory database; if the package is not found there, it falls back to the standard ecosystem advisories.

> When Trivy detects a Seal-patched package by either scheme, it automatically uses Seal Security advisories for vulnerability scanning.

Reviewer: This is related to both the `Renamed` and `NoPrefix` paragraphs.
It makes sense to move this into the `Application Dependencies` section.

Author: Moved into the Application Dependencies intro so it covers both schemes (718a6f5).
